web/interceptor/helpers.go

```diff
@@ -0,0 +1,58 @@
+package interceptor
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/quickfeed/quickfeed/web/auth"
+	"github.com/quickfeed/quickfeed/web/auth/tokens"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// getAuthenticatedContext returns a new context with the user ID attached to it.
+// If the context does not contain a valid session cookie, it returns an error.
+func getAuthenticatedContext(ctx context.Context, logger *zap.SugaredLogger, tm *tokens.TokenManager) (context.Context, error) {
+	meta, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		logger.Error("Failed to extract metadata")
+		return nil, ErrContextMetadata
+	}
+	token, err := extractToken(meta)
+	if err != nil {
+		logger.Errorf("Failed to extract token: %v", err)
+		return nil, ErrInvalidAuthCookie
+	}
+	claims, err := tm.GetClaims(token)
+	if err != nil {
+		logger.Errorf("Failed to extract claims from JWT: %v", err)
+		return nil, ErrInvalidAuthCookie
+	}
+	if tm.UpdateRequired(claims) {
+		logger.Debug("Updating token for user ", claims.UserID)
+		updatedToken, err := tm.NewAuthCookie(claims.UserID)
+		if err != nil {
+			logger.Errorf("Failed to update cookie: %v", err)
```

Comment:
Should we continue after this fail? or should the error be returned?

Reply:
I think it is better to return an error. Fixed.